src/lib.rs

```diff
@@ -234,6 +234,162 @@ use std::sync::{Arc, Condvar, Mutex, MutexGuard, RwLock, TryLockError};
+            $crate::async_eval($name, |_| {
+                panic!("Return is not supported for the fail point \"{}\"", $name);
+            })
+            .await;
+        }};
+        ($name:expr, $e:expr) => {{
+            if let Some(res) = $crate::async_eval($name, $e).await {
+                return res;
+            }
+        }};
+        ($name:expr, $cond:expr, $e:expr) => {{
+            if $cond {
+                $crate::async_fail_point!($name, $e);
+            }
+        }};
+    }
+
+    /// Configures an async callback to be triggered at the specified
+    /// failpoint. If the failpoint is not implemented using
+    /// `async_fail_point`, the execution will raise an exception.
+    pub fn cfg_async_callback<S, F>(name: S, f: F) -> Result<(), String>
+    where
+        S: Into<String>,
+        F: Fn() -> BoxFuture<'static, ()> + Send + Sync + 'static,
+    {
+        let mut registry = REGISTRY.registry.write().unwrap();
+        let p = registry
+            .entry(name.into())
+            .or_insert_with(|| Arc::new(FailPoint::new()));
+        let action = Action::from_async_callback(f);
+        let actions = vec![action];
+        p.set_actions("callback", actions);
+        Ok(())
+    }
+
+    #[doc(hidden)]
+    pub async fn async_eval<R, F: FnOnce(Option<String>) -> R>(name: &str, f: F) -> Option<R> {
+        let p = {
+            let registry = REGISTRY.registry.read().unwrap();
+            match registry.get(name) {
+                None => return None,
+                Some(p) => p.clone(),
+            }
+        };
+        p.async_eval(name).await.map(f)
+    }
+
+    impl Action {
+        #[cfg(feature = "async")]
+        fn from_async_callback(
+            f: impl Fn() -> BoxFuture<'static, ()> + Send + Sync + 'static,
+        ) -> Action {
+            let task = Task::CallbackAsync(AsyncCallback::new(f));
+            Action {
+                task,
+                freq: 1.0,
+                count: None,
+            }
+        }
+    }
+
+    impl FailPoint {
+        #[cfg_attr(feature = "cargo-clippy", allow(clippy::option_option))]
+        async fn async_eval(&self, name: &str) -> Option<Option<String>> {
+            let task = {
+                let task = self
+                    .actions
+                    .read()
+                    .unwrap()
+                    .iter()
+                    .filter_map(Action::get_task)
+                    .next();
+                match task {
+                    Some(Task::Pause) => {
+                        // let n = self.async_pause_notify.clone();
+                        self.async_pause_notify.notified().await;
+                        return None;
+                    }
+                    Some(t) => t,
+                    None => return None,
+                }
+            };
+
+            match task {
+                Task::Off => {}
+                Task::Return(s) => return Some(s),
+                Task::Sleep(_) => panic!(
```

Comment:
I do worry this is not user friendly enough.

Maybe we can spawn a thread and send a signal after sleep some while to implement this without depending on any async runtime.

Reply:
Done